libs/logstream/streamer.go

```diff
@@ -0,0 +1,506 @@
+type Dialer interface {
+	DialContext(ctx context.Context, urlStr string, requestHeader http.Header) (*websocket.Conn, *http.Response, error)
+}
+
+// TokenProvider refreshes tokens when the streamer needs a new bearer token.
+type TokenProvider func(context.Context) (string, error)
+
+// Config holds the options for running a log stream.
+type Config struct {
+	Dialer        Dialer
+	URL           string
+	Origin        string
+	Token         string
+	TokenProvider TokenProvider
+	Search        string
+	Sources       map[string]struct{}
+	Tail          int
+	Follow        bool
+	Prefetch      time.Duration
+	Writer        io.Writer
+	UserAgent     string
+}
+
+// Run connects to the log stream described by cfg and copies frames to the writer.
+func Run(ctx context.Context, cfg Config) error {
+	if cfg.Writer == nil {
+		return errors.New("logstream: writer is required")
+	}
+
+	streamer := &logStreamer{
+		dialer:        cfg.Dialer,
+		url:           cfg.URL,
+		origin:        cfg.Origin,
+		token:         cfg.Token,
+		tokenProvider: cfg.TokenProvider,
+		search:        cfg.Search,
+		sources:       cfg.Sources,
+		tail:          cfg.Tail,
+		follow:        cfg.Follow,
+		prefetch:      cfg.Prefetch,
+		writer:        cfg.Writer,
+		userAgent:     cfg.UserAgent,
+	}
+	if streamer.userAgent == "" {
+		streamer.userAgent = defaultUserAgent
+	}
+	return streamer.Run(ctx)
+}
+
+type logStreamer struct {
+	dialer        Dialer
+	url           string
+	origin        string
+	token         string
+	tokenProvider TokenProvider
+	search        string
+	sources       map[string]struct{}
+	tail          int
+	follow        bool
+	prefetch      time.Duration
+	writer        io.Writer
+	tailFlushed   bool
+	userAgent     string
+}
+
+func (s *logStreamer) Run(ctx context.Context) error {
+	if s.dialer == nil {
+		s.dialer = &websocket.Dialer{}
+	}
+	backoff := 200 * time.Millisecond
+	timer := time.NewTimer(time.Hour)
+	stopTimer(timer)
+	defer timer.Stop()
+	for {
+		if err := s.ensureToken(ctx); err != nil {
+			return err
+		}
+
+		headers := http.Header{}
+		headers.Set("Authorization", "Bearer "+s.token)
+		headers.Set("User-Agent", s.userAgent)
+		if s.origin != "" {
+			headers.Set("Origin", s.origin)
+		}
+
+		conn, resp, err := s.dialer.DialContext(ctx, s.url, headers)
+		if err != nil {
+			err = decorateDialError(err, resp)
+		} else if resp != nil && resp.Body != nil {
+			_ = resp.Body.Close()
+		}
+		if err != nil {
+			if ctx.Err() != nil {
+				return ctx.Err()
+			}
+			if s.follow && s.shouldRefreshForStatus(resp) {
+				if err := s.refreshToken(ctx); err != nil {
+					return err
+				}
+				backoff = time.Second
+				continue
+			}
+			if !s.follow {
+				return err
+			}
+			if err := waitForBackoff(ctx, timer, backoff); err != nil {
+				return err
+			}
+			backoff = min(backoff*2, 5*time.Second)
+			continue
+		}
+
+		backoff = time.Second
+		err = s.consume(ctx, conn)
+		_ = conn.Close()
```

Comment:
The refactoring (splitting the method to multiple ones) would be helpful as then we could follow the best practices and add `defer conn.Close()` right after a successful connection:

```go
        conn, resp, err := s.dialer.DialContext(ctx, s.url, headers)
		if err != nil {
			// err handling
		}
        defer conn.Close() // <-- here
```

and also, within the same function, we could move the goroutine:
```go
	go func() {
		select {
		case <-ctx.Done():
			_ = conn.WriteControl(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, "context canceled"), time.Now().Add(time.Second))
			_ = conn.Close()
		case <-closeCh:
		}
	}()
```

Reply:
Also, currently the connection is closed two times and in case of timeouted context, second call would return an error (connection already closed). Could we use `sync.Once` to solve it?